### PR TITLE
Axe API integration with Jest and Puppeteer

### PR DESCRIPTION
**Pull request by gziolo**
## Description
Implementation based on #12743.

Hopefully this is th final try to include `axe` integration with out e2e tests setup. This PR introduces new `Jest` matcher which takes a `Puppeteer` `page` instance and check if it is accesible as shown below:

```js
await expect( page ).toPassAxeTests( {
	include: '.block-editor',
	exclude: '#metaboxes, #post-title-0, .editor-default-block-appender__content',
} );
```

## How has this been tested?
`npm run test-e2e` with the following patch applied in addition to those changes:

```diff
diff --git a/packages/tests-e2e/support/setup-test-framework.js b/packages/tests-e2e/support/setup-test-framework.js
index cc8168644da..89cfc37b127 100644
--- a/packages/tests-e2e/support/setup-test-framework.js
+++ b/packages/tests-e2e/support/setup-test-framework.js
@@ -162,6 +162,7 @@ beforeAll( async () => {
 } );
 
 afterEach( async () => {
+	await expect( page ).toPassAxeTests();
 	await setupBrowser();
 } );
```

## Types of changes
New feature for e2e testing.

## Known issues

There is a bug in Jest with async matcher which makes them print differently than regular matchers. See https://github.com/facebook/jest/issues/7066 for more details. They are working on a fix and it looks like it might land in the next major release of Jest.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate.

### CODE REVIEW

**Comment by gziolo** [comment]
This is ready for review. This PR in the final version adds only new package which exposes the async matcher to validate whether a given page or its part is accessible according to Axe tool analysis.

@stephenmathieson and @dylanb, I would appreciate your feedback on this one as it's based on `axe-puppeteer` and your great work started at WC US (#12743) :)

File: packages/jest-puppeteer-axe/README.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+```js
+import '@wordpress/jest-puppeteer-axe';
+```
+
+## Usage
+
+In your Jest test suite add the following code to the test's body:
+
+```js
+test( 'checks accessibility of the test page', async () => {
```
Comment (greatislander) [inline packages/jest-puppeteer-axe/README.md]:
```suggestion
test( 'Checks the test page with Axe', async () => {
```

File: packages/jest-puppeteer-axe/README.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
+# Jest Puppeteer Axe
+
+[Axe](https://www.deque.com/axe/) (the Accessibility Engine) API integration with [Jest](https://jestjs.io/) and [Puppeteer](https://pptr.dev/).
+
+Defines Jest async matcher to check whether a given Puppeteer's page instance is accessible.
```
Comment (greatislander) [inline packages/jest-puppeteer-axe/README.md]:
```suggestion
Defines Jest async matcher to check whether a given Puppeteer's page instance passes [Axe](https://www.deque.com/axe/) accessibility tests.
```

File: packages/jest-puppeteer-axe/README.md [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+## Usage
+
+In your Jest test suite add the following code to the test's body:
+
+```js
+test( 'checks accessibility of the test page', async () => {
+	// First, run some code which loads the content of the page.
+	loadTestPage();
+	
+	await expect( page ).toBeAccessible();
```
Comment (greatislander) [inline packages/jest-puppeteer-axe/README.md]:
```suggestion
	await expect( page ).toPassAxeTests();
```

File: packages/jest-puppeteer-axe/README.md [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+} );
+```
+
+It is also possible to pass optional Axe API options to perform customized check:
+- `include` - CSS selector to to add the list of elements to include in analysis.
+- `exclude` - CSS selector to to add the list of elements to exclude from analysis.
+
+```js
+test( 'checks accessibility of the test component excluding some button', async () => {
```
Comment (greatislander) [inline packages/jest-puppeteer-axe/README.md]:
```suggestion
test( 'Checks the test component with Axe excluding some button', async () => {

```

File: packages/jest-puppeteer-axe/README.md [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+	// First, run some code which loads the content of the page.
+	loadPageWithTestComponent();
+	
+	await expect( page ).toBeAccessible( {
```
Comment (greatislander) [inline packages/jest-puppeteer-axe/README.md]:
```suggestion
	await expect( page ).toPassAxeTests( {
```

File: packages/jest-puppeteer-axe/src/index.js [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+					output += `      - ${ item.message }.\n`;
+				} );
+			}
+		} );
+		return output;
+	} ).join( '\n' );
+}
+
+/**
+ * Defines async matcher to check whether a given Puppeteer's page instance is accessible.
```
Comment (greatislander) [inline packages/jest-puppeteer-axe/src/index.js]:
```suggestion
 * Defines async matcher to check whether a given Puppeteer's page instance passes Axe accessibility tests.
 *
 * @see https://www.deque.com/axe/
```

File: packages/jest-puppeteer-axe/src/index.js [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+ * @param {Page}    page             Puppeteer's page instance.
+ * @param {?Object} params           Optional Axe API options.
+ * @param {?string} params.include   CSS selector to add to the list of elements
+ *                                   to include in analysis.
+ * @param {?string} params.exclude   CSS selector to add to the list of elements
+ *                                   to exclude from analysis.
+ *
+ * @return {Object} A matcher object with two keys `pass` and `message`.
+ */
+async function toBeAccessible( page, { include, exclude } = {} ) {
```
Comment (greatislander) [inline packages/jest-puppeteer-axe/src/index.js]:
```suggestion
async function toPassAxeTests( page, { include, exclude } = {} ) {
```

File: packages/jest-puppeteer-axe/src/index.js [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+	if ( exclude ) {
+		axe.exclude( exclude );
+	}
+
+	const { violations } = await axe.analyze();
+
+	const pass = violations.length === 0;
+	const message = pass ?
+		() => {
+			return this.utils.matcherHint( '.not.toBeAccessible' ) +
```
Comment (greatislander) [inline packages/jest-puppeteer-axe/src/index.js]:
```suggestion
			return this.utils.matcherHint( '.not.toPassAxeTests' ) +
```

File: packages/jest-puppeteer-axe/src/index.js [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+				'\n\n' +
+				'Expected page to contain accessibility check violations.\n' +
+				'No violations found.';
+		} :
+		() => {
+			return this.utils.matcherHint( '.toBeAccessible' ) +
```
Comment (greatislander) [inline packages/jest-puppeteer-axe/src/index.js]:
```suggestion
			return this.utils.matcherHint( '. toPassAxeTests' ) +
```

File: packages/jest-puppeteer-axe/src/index.js [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+				'\n\n' +
+				'Expected page to be accessible.\n' +
```
Comment (greatislander) [inline packages/jest-puppeteer-axe/src/index.js]:
```suggestion
				'Expected page to pass Axe accessibility tests.\n' +
```

File: packages/jest-puppeteer-axe/src/index.js [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+		};
+
+	return {
+		message,
+		pass,
+	};
+}
+
+expect.extend( {
+	toBeAccessible,
```
Comment (greatislander) [inline packages/jest-puppeteer-axe/src/index.js]:
```suggestion
	toPassAxeTests,
```

**Review by greatislander — CHANGES_REQUESTED** [review]
I've proposed some changes based on concerns that the function name and documentation could lead users to think that running Axe alone can accurately determine whether or not a page is accessible. Any automated accessibility tool can only catch a certain subset of potential accessibility issues, and must be combined with manual evaluation. It is probably better not to use language like:

```javascript
expect( page ).toBeAccessible();
```

But rather to use documentation and function names which indicates that the evaluation is based on criteria of this specific tool, e.g.:

```javascript
expect( page ).toPassAxeTests();
```

Open to and grateful for other ideas of how to frame this.

**Comment by gziolo** [comment]
> I've proposed some changes based on concerns that the function name and documentation could lead users to think that running Axe alone can accurately determine whether or not a page is accessible. Any automated accessibility tool can only catch a certain subset of potential accessibility issues, and must be combined with manual evaluation.

Totally agree with your reasoning 👍 I didn't think about it at all. I was too focused on the technical aspects. Thanks for your review. I applied all the changes you proposed as they better explain the intents of this new Jest matcher introduced. I'm not 100% sure if `tests` in `expect( page ).toPassAxeTests();` is the best fit but I'm having hard time finding out a better alternative. Some ideas I had are:
```js
expect( page ).toPassAxeChecks();
expect( page ).toPassAxeAnalysis();
expect( page ).toPassAxeVerification();
```
We can totally proceed as is if there isn't a better proposal.

**Comment by greatislander** [comment]
@gziolo thanks for accepting the suggestions! I hear the concern around using `Tests` in the function name. Of your suggestions I think I'd lean towards:

```
expect( page ).toPassAxeChecks();
```

I also thought of:

```
expect( page ).toPassAxeAudit();
```

Open to other suggestions too.

File: packages/tests-e2e/support/setup-test-framework.js [hunk 12]
```diff
@@ -8,6 +8,7 @@ import { get } from 'lodash';
  * WordPress dependencies
  */
 import '@wordpress/jest-console';
+import '@wordpress/jest-puppeteer-axe';
```
Comment (aduth) [inline packages/tests-e2e/support/setup-test-framework.js]: If we're not using it, questionable to be opting in now, vs. PR to be about only adding the new module. Can be fine to limit the size of the follow-up pull request, having no/minimal impact here.

**Review by aduth — APPROVED** [review]
The messaging of a failing test is non-conventional, with only an "Error" text being highlighted. In direct conversation, it was mentioned this is a bug present in Jest for the moment. Pending @gziolo to share.

In the meantime, this appears good as it stands. 👍

File: packages/tests-e2e/support/setup-test-framework.js (hunk 12 above)
Comment (gziolo) [reply packages/tests-e2e/support/setup-test-framework.js]: yes, will remove for now

**Comment by dylanb** [comment]
I am a little late on this but this looks like great progress

**Comment by dylanb** [comment]
@gziolo seems like a good improvement would be to publish a `axe-jest-puppeteer` module that implements the matcher and use that - WDYT?

cc/ @wilcofiers @stephenmathieson

**Comment by gziolo** [comment]
This is going to be published to npm as `@wordpress/jest-puppeteer-axe` closer to the end of February, is it what you mean?

**Pull request merged**